Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp

```diff
@@ -63,6 +63,56 @@ static inline bool matches_lang_pseudo_class(DOM::Element const& element, Vector
     return false;
 }
 
+// https://drafts.csswg.org/selectors-4/#relational
+static inline bool matches_has_pseudo_class(CSS::Selector const& selector, Optional<CSS::CSSStyleSheet const&> style_sheet_for_rule, DOM::Element const& element)
+{
+    switch (selector.compound_selectors()[0].combinator) {
+    // Shouldn't be possible because we've parsed relative selectors, which always have a combinator, implicitly or explicitly.
+    case CSS::Selector::Combinator::None:
+        VERIFY_NOT_REACHED();
+    case CSS::Selector::Combinator::Descendant: {
+        bool has = false;
+        element.for_each_in_subtree([&](auto const& descendant) {
+            if (!descendant.is_element())
+                return TraversalDecision::Continue;
+            auto const& descendant_element = static_cast<DOM::Element const&>(descendant);
+            // Provide the element as our "scope" (which is referred to as the "anchor element" in the spec: https://drafts.csswg.org/selectors-4/#relative-selector)
+            if (matches(selector, style_sheet_for_rule, descendant_element, {}, element)) {
```

Comment:
Not sure if using the `scope` parameter is okay here. `scope` is meant for scoped selectors, but I'm using it here for relative selectors. The two technically never conflict, but I'm using one parameter for two different use-cases, which might be a little weird.

The other option is adding yet another parameter to the `matches*` family of functions (an `Optional<DOM::Element const&>` representing the anchor for the relative selector). Regardless, I feel like there's a better way to implement this.

Reply:
Removed this as of latest force push.